drivers/soundwire/intel_init.c

```diff
@@ -56,26 +94,24 @@ static bool is_link_enabled(struct fwnode_handle *fw_node, int i)
 
 static int sdw_intel_cleanup(struct sdw_intel_ctx *ctx)
 {
-	struct sdw_intel_link_res *link = ctx->links;
+	struct sdw_intel_link_dev *ldev = ctx->ldev;
 	u32 link_mask;
 	int i;
 
-	if (!link)
+	if (!ldev)
 		return 0;
 
 	link_mask = ctx->link_mask;
 
-	for (i = 0; i < ctx->count; i++, link++) {
+	for (i = 0; i < ctx->count; i++, ldev++) {
 		if (!(link_mask & BIT(i)))
 			continue;
 
-		if (link->pdev) {
-			pm_runtime_disable(&link->pdev->dev);
-			platform_device_unregister(link->pdev);
-		}
+		pm_runtime_disable(&ldev->auxdev.dev);
+		intel_link_dev_unregister(ldev);
```

Comment:
This function is also called on the error path in `sdw_intel_probe_controller()` in which case only some of the link devices from the mask might have initialised successfully. Shouldn't this loop check that? Maybe we could use the `ctx->link_list` here? And BTW, don't we have to remove the device from the list?

Reply:
I completely reworked the error flow to deal with failures on device init/add stages, and also removed devm_ since we may call the allocation from a workqueue. I am not sure everything's fully correct now but it looks cleaner.